### PR TITLE
feat: support scaling for fragments with no shuffle dispatcher

### PR DESCRIPTION
**Pull request by shanicky**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?
This PR provides scaling ablities for Fragments upstream and downstream of NoShuffle dispatcher

The current implementation iterates through the table fragments when building the reschedule context, which detects the type of the dispatcher and creates an index of the no shuffle fragment.

For fragments downstream of NoShuffle, when scaling
1. no rebalance vnode operation is performed on them, all vnode assignments are copied directly from the root of the NoShuffle cascade relationship, and only the root of the NoShuffle is rebalanced.
2. skip the creation of HangingChannel, because the upstream and downstream of NoShuffle expand and shrink at the same time, there is no HangingChannel, unlike Hash, which needs to draw HangingChannel from the rest of the upstreams
3. Skip `upstream_fragment_dispatcher_ids` and `upstream_dispatcher_mapping` in the reschedule, because there is no upstream to update.

For the fragment upstream of NoShuffle
1. the `downstream_fragment_id` in the reschedule should always be None, because there is no downstream `MergeNode` to update
2. we need to modify the downstream actor id in the dispatcher, where the length should always be 1.

This PR test uses the following SQL, which creates a diamond dependency

```
create table t1 (v1 int);
create materialized view m1 as select * from t1 where v1 > 10;
create materialized view m2 as select * from t1 where v1 > 20;
create materialized view m3 as select * from m2 where v1 < 70;
create materialized view m4 as select m1.v1 as m1v, m3.v1 as m3v from m1 join m3 on m1.v1 = m3.v1;
create materialized view m5 as select * from m3;
```

Then, after inserting 1 to 100 into t1 and continuously reschedule, the data in m5 should always be 21 to 69 with 49 numbers.

The specific DAG looks like this

image

This PR also detects a bug, the `upstream_actor_id` of the chain actor are not set.

## Checklist

- [x] I have written necessary rustdoc comments
~- [ ] I have added necessary unit tests and integration tests~
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)

https://github.com/risingwavelabs/risingwave/issues/5740

### CODE REVIEW

**Comment by BugenZhao** [comment]
Cool! 🚀🚀🚀 How about adding the test to `src/tests/simulation_scale/tests/mv_on_mv.rs` so that we can test it automatically with madsim? We may take `nexmark_q4_materialize_agg_cache_invalidation` as an example.

**Comment by shanicky** [comment]
> Cool! 🚀🚀🚀 How about adding the test to `src/tests/simulation_scale/tests/mv_on_mv.rs` so that we can test it automatically with madsim? We may take `nexmark_q4_materialize_agg_cache_invalidation` as an example.

ok, i'll add that

File: src/meta/src/stream/scale.rs [hunk 1]
```diff
@@ -549,39 +728,46 @@ where
...
-                                host: None,
-                            }),
-                            downstream: Some(ActorInfo {
-                                actor_id: *new_actor_id,
-                                host: worker.host.clone(),
-                            }),
-                        });
+
+                // for chain (no shuffle) downstreams, we don't need to create hanging channels
+                if !ctx.no_shuffle_target_fragment_ids.contains(fragment_id) {
```
Comment (BugenZhao) [inline src/meta/src/stream/scale.rs:733]: How about asserting `upstream_worker_id == worker.id` here?

**Review by BugenZhao — COMMENTED** [review]
Generally LGTM. Could you please add more comments for the logic, as it seems more and more complicated? 🤯

File: src/meta/src/stream/scale.rs [hunk 2]
```diff
@@ -522,7 +572,136 @@ where
...
+                no_shuffle_upstream_actor_map.insert(*actor_id as ActorId, upstream_actor_id);
+                no_shuffle_downstream_actors_map
+                    .entry(upstream_actor_id)
+                    .or_default()
+                    .insert(*fragment_id, *actor_id);
+            }
+
+            assert!(fragment_updated_bitmap
+                .insert(*fragment_id, fragment_bitmap)
+                .is_none());
```
[the remark this answers is not in the document]
Comment (shanicky) [reply src/meta/src/stream/scale.rs:666]: This place needs to exclude the diamond dependency, so it should be asserted that it has not been inserted before

File: src/meta/src/stream/scale.rs (hunk 1 above)
Comment (shanicky) [reply src/meta/src/stream/scale.rs:733]: This index is created at the beginning based on the status of the existing actor, so for the chain fragment's upstream will not be found in the follow `actor_id_to_parallel_unit`

**Review by yezizp2012 — COMMENTED** [review]
Generally LGTM, great job!!!

File: src/meta/src/stream/scale.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shanicky) [reply src/meta/src/stream/scale.rs]: use try_insert and unwrap ed

**Comment by shanicky** [comment]
Added a simple simulation test on noshuffle relation

**Review by yezizp2012 — APPROVED** [review]
Great job!! LGTM~

File: src/tests/simulation_scale/src/ctl_ext.rs [hunk 3]
```diff
@@ -74,6 +78,17 @@ pub mod predicate {
         Box::new(p)
     }
 
+    /// There exists operators whose identity contains `s` in the fragment.
```
Comment (BugenZhao) [inline src/tests/simulation_scale/src/ctl_ext.rs:81]: May also update the doc?

**Review by BugenZhao — APPROVED** [review]
LGTM!

File: src/tests/simulation_scale/src/ctl_ext.rs (hunk 3 above)
Comment (BugenZhao) [reply src/tests/simulation_scale/src/ctl_ext.rs:81]:
```suggestion
    /// There does not exist any operator whose identity contains `s` in the fragment.
```

**Pull request merged**